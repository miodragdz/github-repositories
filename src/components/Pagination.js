import React, { Component } from 'react'
import { range } from '../utils/helpers'
import { LEFT_PAGE, RIGHT_PAGE } from '../utils/constants'

export class Pagination extends Component {
  getPageNumbers = () => {
    const { currentPage, maxPage } = this.props
    const startPage = Math.max(2, currentPage - 2);
    const endPage = Math.min(maxPage - 1, currentPage + 2);

    let pages = range(startPage, endPage);

    const hasLeftSpill = startPage > 2;              //has hidden pages to the left
    const hasRightSpill = (maxPage - endPage) > 1;   //has hidden pages to the right
    const spillOffset = 7 - (pages.length + 1);      //number of hidden pages either to the left or to the right

    // handle: (1) < {5 6} [7] {8 9} (10)
    if (hasLeftSpill && !hasRightSpill) {
      const extraPages = range(startPage - spillOffset + 1, startPage - 1);
      pages = [LEFT_PAGE, ...extraPages, ...pages];
      return [1, ...pages, maxPage];
    }

    // handle: (1) {2 3} [4] {5 6} > (10)
    if (!hasLeftSpill && hasRightSpill) {
      const extraPages = range(endPage + 1, endPage + spillOffset - 1);
      pages = [...pages, ...extraPages, RIGHT_PAGE];
      return [1, ...pages, maxPage];
    }

    // handle: (1) < {4 5} [6] {7 8} > (10)
    pages = [LEFT_PAGE, ...pages, RIGHT_PAGE];
    return [1, ...pages, maxPage];
  }
  
  render() {
    const { jumpToPage, currentPage, maxPage } = this.props
    const pages = this.getPageNumbers();
  
    return (
      <div className="pagination">
        {pages.map((page) => {
          if (page === LEFT_PAGE) return (
              <button key={page} className="prev page-numbers" aria-label="Previous" onClick={() => jumpToPage(currentPage-5) }>
                <span>&laquo;</span>
                <span>Previous</span>
              </button>
          );

          if (page === RIGHT_PAGE) return (
              <button key={page} className="next page-numbers" aria-label="Next" onClick={() => jumpToPage(currentPage+5) }>
                <span>Next</span>
                <span>&raquo;</span>             
              </button>
          );

          return (
            <button 
              key={page} 
              className={`page-numbers${ currentPage === page ? ' current' : ''}`}  
              onClick={ () => jumpToPage(Math.min(page, maxPage)) }
            >
              {page}
            </button>
          );
        })}
      </div>
    )
  }
}

export default Pagination
