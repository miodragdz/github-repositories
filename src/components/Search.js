import React, { Component } from "react";

export class Search extends Component {
  state = {
    searchTerm: '',
  };

  onHandleChange = e => {
    this.setState({
      searchTerm: e.target.value
    });
  }

  onHandleSubmit = e => {
    const { onSearchButtonClick } = this.props;
    const { searchTerm } = this.state;
    e.preventDefault();
    onSearchButtonClick(searchTerm);
  }

  render() {   
    const { searchTerm } = this.state;

    return (
      <div className="search">
        <form className="form">
          <input
            id="location"
            value={searchTerm}
            placeholder="Search by name"
            onChange={this.onHandleChange}
            className="form__input"
          />
          <button className="form__button" disabled={!searchTerm} onClick={this.onHandleSubmit} type="submit">Search</button>        
        </form>
      </div>
    );
  }
}

export default Search;
