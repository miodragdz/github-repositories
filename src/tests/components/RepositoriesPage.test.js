import React from "react";
import { shallow, mount } from "enzyme";
import configureMockStore from "redux-mock-store";
import thunk from "redux-thunk";
import RepositoriesPageContainer, {
  RepositoriesPage,
} from "../../components/RepositoriesPage";
import Pagination from "../../components/Pagination";
import Search from "../../components/Search";
import { repositories, noDataFatched } from "../fixtures/repositories";
import actionFunctions from "../../actions/repositoriesAction";

const createMockStore = configureMockStore([thunk]);

test("should render RepositoriesPage with repositories", () => {
  const wrapper = shallow(
    <RepositoriesPage
      getRepositories={jest.fn(() => Promise.resolve(repositories))}
      repositories={repositories.items}
      totalCount={repositories.total_count}
    />
  );

  expect(wrapper).toMatchSnapshot();
});

test("should render RepositoriesPage with message", async () => {
  const wrapper = shallow(
    <RepositoriesPage
      getRepositories={jest.fn(() => Promise.resolve(noDataFatched))}
      repositories={[]}
      totalCount={0}
    />
  );

  await wrapper.find(Search).props().onSearchButtonClick("zxcvbnmlkgf");

  expect(wrapper).toMatchSnapshot();
  expect(wrapper.find("h2")).toHaveLength(2);
});

test("should change state on invoke onSearchButtonClick callback", async () => {
  const wrapper = shallow(
    <RepositoriesPage
      getRepositories={jest.fn(() => Promise.resolve(repositories))}
      repositories={repositories.items}
      totalCount={repositories.total_count}
    />
  );

  await wrapper.find(Search).props().onSearchButtonClick("jquery");

  expect(wrapper.state("searchTerm")).toEqual("jquery");
  expect(wrapper.state("maxPage")).toEqual(40);
  expect(wrapper.state("currentPage")).toEqual(1);
});

test("should change state on invoke jumpToPage callback", async () => {
  const wrapper = shallow(
    <RepositoriesPage
      getRepositories={jest.fn(() => Promise.resolve(repositories))}
      repositories={repositories.items}
      totalCount={repositories.total_count}
    />
  );
  wrapper.setState({ maxPage: 40 });

  await wrapper.find(Pagination).props().jumpToPage(3);

  expect(wrapper.state("currentPage")).toEqual(3);
});

test("mapStateToProps - should provide correct props from store", () => {
  const state = { repositories };
  const mockStore = createMockStore(state);
  const wrapper = shallow(<RepositoriesPageContainer store={mockStore} />);

  expect(wrapper.props().children.props.totalCount).toEqual(
    repositories.total_count
  );
  expect(wrapper.props().children.props.repositories).toEqual(
    repositories.items
  );
});

test("mapDispatchToProps - should provide dispatch from store", async () => {
  const state = { repositories };
  const mockStore = createMockStore(state);

  actionFunctions.getRepositories = jest
    .fn()
    .mockImplementation((dispatch, searchTerm, pageNumber) => {
      return Promise.resolve(repositories).then((repositories) => {
        dispatch(actionFunctions.setRepositories(repositories));
        return repositories;
      });
    });

  const wrapper = mount(<RepositoriesPageContainer store={mockStore} />);

  await wrapper.find(Search).props().onSearchButtonClick("jquery", 1);

  const actions = mockStore.getActions();
  expect(actions).toHaveLength(1);
});
