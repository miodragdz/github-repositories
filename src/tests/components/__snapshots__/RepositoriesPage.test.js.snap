// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render RepositoriesPage with message 1`] = `ShallowWrapper {}`;

exports[`should render RepositoriesPage with repositories 1`] = `ShallowWrapper {}`;
